### ISSUE TITLE
Management-center could not lock home directory

### ISSUE THREAD

**Issue by jperucha** [hazelcast/charts#140]
Hi,

I'm deploying the Hazelcast Helm chart version 3.4.3 (https://hazelcast-charts.s3.amazonaws.com/) on an EKS Kubernetes cluster.

To avoid another issue related to adminCredentialsSecretName helm chart property I have manually created a pvc and indicated in the values of mancenter to enable persistence and use that pvc.
The first deployment of man center is working without problems, it uses that pvc and binds the pod to it and when I enter it works. 
The issue occurs on the following times the pod is created, as the config is stored in the pv created by the pvc, this error appears and man center never starts again.

`using automatic sizing of heap size by up to 80% of available memory and starting with container support
executing command specified by MC_INIT_CMD for container initialization
ERROR: Could not lock home directory. Make sure that Management Center web application is stopped (offline) before starting this command. If you are sure the application is stopped, it means that lock file was not deleted properly. Please delete 'mc.lock' file in the home directory manually before using the command.
To see the full stack trace, re-run with the -v/--verbose option.`

Said that, here below I provide the config values applied to the management center deployment in this chart.
![image](https://user-images.githubusercontent.com/11365383/87288962-54b25f00-c4fc-11ea-9400-22f2dff56a4f.png)

Ingress host and tls values are also configured but omitted here.

This is the manually created pvc config
![image](https://user-images.githubusercontent.com/11365383/87289433-ea4dee80-c4fc-11ea-8132-77566adf6c08.png)

Storage class (default) is an EBS GP2

Seems like a command to delete mc.lock file is needed as man center is not working, therefore I consider it stopped.

**Comment by leszko** [comment]
@jperucha could you check if the issue persists with the latest Helm Chart version `3.4.10`?

**Comment by jperucha** [comment]
Hi @leszko.
The latest hazelcast chart version available to use is 3.4.9, so i have tried with that version. I see that Hazelcast-enterprise is the chart that has 3.4.10 version available, maybe hazelcast chart is outdated compared to hazelcast-enterprise.
  
Either way, the result is the same. I imagine logging was improved as it is now more verbose, but the error is still the same as it was 2 months ago. Check out the traze below.

```
using automatic sizing of heap size by up to 80% of available memory and starting with container support
executing command specified by MC_INIT_CMD for container initialization
2020-09-23 14:20:17 [main] INFO  c.h.webmonitor.config.BuildInfo - Hazelcast Management Center version 4.2020.08, Jet client version: 4.2, IMDG client version: 4.0.1, embedded Jetty version: 9.4.30.v20200611
2020-09-23 14:20:17 [main] INFO  c.h.webmonitor.config.BuildInfo - This version of Hazelcast Management Center is tested with the following cluster minor versions: IMDG 4.0, Jet 4.2
ERROR: Could not lock home directory. Make sure that Management Center web application is stopped (offline) before starting this command. If you are sure the application is stopped, it means that lock file was not deleted properly. Please delete 'mc.lock' file in the home directory manually before using the command.
To see the full stack trace, re-run with the -v/--verbose option.
```

**Comment by leszko** [comment]
@hasancelik could you check this?

**Comment by martinmares** [comment]
Helm chart 3.5.0 is affected with same "mystic" error:

```
using automatic sizing of heap size by up to 80% of available memory and starting with container support
executing command specified by MC_INIT_CMD for container initialization
2020-11-25 22:37:00 [main] INFO  c.h.webmonitor.config.BuildInfo - Hazelcast Management Center 4.2020.10 (20201029 - a78b68f), Jet client version: 4.3, Hazelcast client version: 4.0.3, embedded Jetty version: 9.4.31.v20200723
ERROR: Could not lock home directory. Make sure that Management Center web application is stopped (offline) before starting this command. If you are sure the application is stopped, it means that lock file was not deleted properly. Please delete 'mc.lock' file in the home directory manually before using the command.
To see the full stack trace, re-run with the -v/--verbose option.
stream closed
```
Try downgrading to chart version 3.4.14 same error:

```
using automatic sizing of heap size by up to 80% of available memory and starting with container support
executing command specified by MC_INIT_CMD for container initialization
2020-11-25 22:48:35 [main] INFO  c.h.webmonitor.config.BuildInfo - Hazelcast Management Center version 4.2020.08, Jet client version: 4.2, IMDG client version: 4.0.1, embedded Jetty version: 9.4.30.v20200611
2020-11-25 22:48:35 [main] INFO  c.h.webmonitor.config.BuildInfo - This version of Hazelcast Management Center is tested with the following cluster minor versions: IMDG 4.0, Jet 4.2
ERROR: Could not lock home directory. Make sure that Management Center web application is stopped (offline) before starting this command. If you are sure the application is stopped, it means that lock file was not deleted properly. Please delete 'mc.lock' file in the home directory manually before using the command.
To see the full stack trace, re-run with the -v/--verbose option.
stream closed
```

Try downgrading to chart version 3.2.1 another error:

```

executing command specified by MC_INIT_CMD
ERROR: Could not lock home directory. Make sure that Management Center web application is stopped (offline) before starting this command. If you are sure the application is stopped, it means that lock file was not deleted properly. Please delete 'mc.lock' file in the home directory manually before using the command.
To see the full stack trace, re-run with the -v/--verbose option.2020-11-25T23:53:20.915783974+01:00 
stream closed
```
**Problem is propably with this PVC storage, which has unnecessary file "mc.lock":**

```
❯ kubectl -n hazelcast get pvc
NAME                                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS           AGE
mancenter-storage-hazelcast-mancenter-0   Bound    pvc-d9c65d89-9322-4e27-bbe2-4415e5ce3adf   8Gi        RWO            rook-ceph-block-ext4   128d
```

After delete this PVC:

```
❯ kubectl -n hazelcast delete pvc mancenter-storage-hazelcast-mancenter-0
```
And installing helm chart version 3.5.0

**Everything is ok:** 

```

using automatic sizing of heap size by up to 80% of available memory and starting with container support
executing command specified by MC_INIT_CMD for container initialization
2020-11-25 23:08:06 [main] INFO  c.h.webmonitor.config.BuildInfo - Hazelcast Management Center 4.2020.10 (20201029 - a78b68f), Jet client version: 4.3, Hazelcast client version: 4.0.3, embedded Jetty version: 9.4.31.v20200723
2020-11-25 23:08:07 [main] INFO  c.h.webmonitor.config.SqlDbConfig - Checking DB for required migrations.
2020-11-25 23:08:08 [main] INFO  c.h.webmonitor.config.SqlDbConfig - Number of applied DB migrations: 0.
WARNING: Cluster null already exists.
To see the full stack trace, re-run with the -v/--verbose option.
##################################################
# initialisation complete, starting now....
##################################################
+ exec java --add-opens java.base/java.lang=ALL-UNNAMED -server -Dhazelcast.mc.home=/data -Djava.net.preferIPv4Stack=true -Dhazelcast.mc.healthCheck.enable=true -DserviceName=hazelcast -Dnamespace=hazelcast -Dhazelcast.mc.tls.enabled=false -XX:+UseContainerSupport -XX:MaxRAMPercentage=80 -cp /opt/hazelcast/management-center/hazelcast-management-center-4.2020.10.jar -Dhazelcast.mc.contextPath=/ -Dhazelcast.mc.http.port=8080 -Dhazelcast.mc.https.port=8443 com.hazelcast.webmonitor.Launcher
2020-11-25 23:08:09 [main] INFO  c.h.webmonitor.config.BuildInfo - Hazelcast Management Center 4.2020.10 (20201029 - a78b68f), Jet client version: 4.3, Hazelcast client version: 4.0.3, embedded Jetty version: 9.4.31.v20200723
2020-11-25 23:08:09 [main] INFO  com.hazelcast.webmonitor.Launcher - Health check is enabled and available at http://localhost:8081/health
2020-11-25 23:08:14 [main] INFO  c.h.webmonitor.config.SqlDbConfig - Checking DB for required migrations.
2020-11-25 23:08:15 [main] INFO  c.h.webmonitor.config.SqlDbConfig - Number of applied DB migrations: 0.
2020-11-25 23:08:15 [main] INFO  c.h.w.s.s.impl.DisableLoginStrategy - Login will be disabled for 5 seconds after 3 failed login attempts. For every 3 consecutive failed login attempts, disable period will be multiplied by 10.
2020-11-25 23:08:15 [main] INFO  c.h.w.config.MetricsStorageConfig - Initializing embedded metrics storage.
2020-11-25 23:08:20 [main] INFO  c.h.w.service.ClusterManager - Connecting to 1 enabled cluster(s) on startup.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.hazelcast.internal.networking.nio.SelectorOptimizer (file:/opt/hazelcast/management-center/hazelcast-management-center-4.2020.10.jar) to field sun.nio.ch.SelectorImpl.selectedKeys
WARNING: Please consider reporting this to the maintainers of com.hazelcast.internal.networking.nio.SelectorOptimizer
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release2020-11-26T00:08:21.601378115+01:00 
2020-11-25 23:08:22 [MC-Client-dev.lifecycle-1] INFO  c.h.w.service.MCClientManager - MC Client connected to cluster dev.
2020-11-25 23:08:22 [MC-Client-dev.event-2] INFO  c.h.w.service.MCClientManager - Started communication with member: Member [10.244.67.253]:5701 - 04e4bb7a-7670-47fd-9cf0-5ba7a92fe546
2020-11-25 23:08:22 [MC-Client-dev.event-2] INFO  c.h.w.service.MCClientManager - Started communication with member: Member [10.244.225.236]:5701 - 37f29264-631d-4a22-862e-28b448299400
2020-11-25 23:08:22 [MC-Client-dev.event-2] INFO  c.h.w.service.MCClientManager - Started communication with member: Member [10.244.82.53]:5701 - 2159eaf8-b0c8-4c7e-9f29-9e0fd79eaf64
2020-11-25 23:08:23 [ClientBwListBroadcaster] INFO  c.h.w.s.ClientBwListBroadcaster - Deploying client filtering config to Member [10.244.225.236]:5701 - 37f29264-631d-4a22-862e-28b448299400 in cluster dev.
2020-11-25 23:08:23 [AsyncExecutor-2] INFO  c.h.w.service.ClusterMetadataStore - Determined type of cluster dev as IMDG.
2020-11-25 23:08:24 [main] INFO  com.hazelcast.webmonitor.Launcher - 
Hazelcast Management Center successfully started at http://localhost:8080/
2020-11-25 23:08:24 [ClientBwListBroadcaster] INFO  c.h.w.s.ClientBwListBroadcaster - Deploying client filtering config to Member [10.244.82.53]:5701 - 2159eaf8-b0c8-4c7e-9f29-9e0fd79eaf64 in cluster dev.
2020-11-25 23:08:24 [ClientBwListBroadcaster] INFO  c.h.w.s.ClientBwListBroadcaster - Deploying client filtering config to Member [10.244.67.253]:5701 - 04e4bb7a-7670-47fd-9cf0-5ba7a92fe546 in cluster dev.
```

```
❯ kubectl -n hazelcast get pods
NAME                    READY   STATUS    RESTARTS   AGE
hazelcast-0             1/1     Running   0          4m25s
hazelcast-1             1/1     Running   0          5m3s
hazelcast-2             1/1     Running   0          5m45s
hazelcast-mancenter-0   1/1     Running   0          3m21s
```

**Comment by leszko** [comment]
Thanks for the comment @martinmares .

@jperucha does it solve your issue?

**Comment by jperucha** [comment]
@leszko The issue is triggered whenever k8s has to recreate mancenter pod for whatever reason, there is no need for human interaction (f.e, chart upgrading) for the issue to appear.
The case @martinmares presents only shows that deleting the pvc will make the chart installation or upgrade work, but it does not show that mancenter pod recreation is going to work if the PVC is not deleted. And to do so, human interaction has to be present to delete it.
IMO, the fix should not require deleting the PVC as it would not make any sense to have a volume dedicated to store information if in the case the pod is erased and has to be recreated, the data in the volume bound by the PVC is deleted in order to have mancenter pod working.
In order to consider the issue fixed, mancenter pod should be able to recreate itself without throwing the error. If any kind of human interaction is needed, the issue is not solved then.

**Comment by martinmares** [comment]
> In order to consider the issue fixed, mancenter pod should be able to recreate itself without throwing the error. If any kind of human interaction is needed, the issue is not solved then.

Yes, this is true, "my solution" is only workaround. The script should do this automatically, if I upgrading helm chart.

**Comment by hasancelik** [comment]
Hi @jperucha and @martinmares ,

We have already created related issue at Management Center repo, as you can see above. This enhancement probably solve the issue you encountered. But I could not reproduce the issue at EKS and GKE with `hazelcast/hazelcast 3.5.0`, here is the my steps for EKS:

1. Create PVC manually:
```
$ cat mc-eks-pvc.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mc-pvc
spec:
  storageClassName: gp2
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi

$ kubectl apply -f  mc-eks-pvc.yaml
persistentvolumeclaim/mc-pvc created

$ kubectl get pvc
NAME     STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
mc-pvc   Pending                                                                                       gp2            8s
```
2. Deploy the chart:
```
$ helm install hz-mc-pvc --set mancenter.persistence.existingClaim=mc-pvc hazelcast/hazelcast --version 3.5.0

$ kubectl get pods
NAME                              READY   STATUS    RESTARTS   AGE
hz-mc-pvc-hazelcast-0             1/1     Running   0          8m25s
hz-mc-pvc-hazelcast-1             1/1     Running   0          7m47s
hz-mc-pvc-hazelcast-2             1/1     Running   0          7m13s
hz-mc-pvc-hazelcast-mancenter-0   1/1     Running   0          8m25s

$ kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM            STORAGECLASS   REASON   AGE
pvc-cf777c4f-8500-4c84-8b06-5634e9630094   1Gi        RWO            Delete           Bound    default/mc-pvc   gp2                     21s
```

3. Restart MC pod:
```
$ kubectl scale statefulset.apps/hz-mc-pvc-hazelcast-mancenter --replicas=0
statefulset.apps/hz-mc-pvc-hazelcast-mancenter scaled

$ kubectl get pods
NAME                              READY   STATUS        RESTARTS   AGE
hz-mc-pvc-hazelcast-0             1/1     Running       0          28m
hz-mc-pvc-hazelcast-1             1/1     Running       0          27m
hz-mc-pvc-hazelcast-2             1/1     Running       0          26m
hz-mc-pvc-hazelcast-mancenter-0   0/1     Terminating   0          28m

$ kubectl scale statefulset.apps/hz-mc-pvc-hazelcast-mancenter --replicas=1
statefulset.apps/hz-mc-pvc-hazelcast-mancenter scaled

$ kubectl get pods
NAME                              READY   STATUS    RESTARTS   AGE
hz-mc-pvc-hazelcast-0             1/1     Running   0          30m
hz-mc-pvc-hazelcast-1             1/1     Running   0          29m
hz-mc-pvc-hazelcast-2             1/1     Running   0          28m
hz-mc-pvc-hazelcast-mancenter-0   1/1     Running   0          70s

$ kubectl logs hz-mc-pvc-hazelcast-mancenter-0
....
Hazelcast Management Center successfully started at http://localhost:8080/
...

$ kubectl describe pod hz-mc-pvc-hazelcast-mancenter-0
Name:               hz-mc-pvc-hazelcast-mancenter-0
...
Events:
  Type    Reason                  Age    From                                                  Message
  ----    ------                  ----   ----                                                  -------
  Normal  Scheduled               5m32s  default-scheduler                                     Successfully assigned default/hz-mc-pvc-hazelcast-mancenter-0 to ip-192-168-17-15.us-west-1.compute.internal
  Normal  SuccessfulAttachVolume  5m29s  attachdetach-controller                               AttachVolume.Attach succeeded for volume "pvc-cf777c4f-8500-4c84-8b06-5634e9630094"
  Normal  Pulled                  5m22s  kubelet, ip-192-168-17-15.us-west-1.compute.internal  Container image "hazelcast/management-center:4.2020.10" already present on machine
  Normal  Created                 5m22s  kubelet, ip-192-168-17-15.us-west-1.compute.internal  Created container hz-mc-pvc-hazelcast-mancenter
  Normal  Started                 5m22s  kubelet, ip-192-168-17-15.us-west-1.compute.internal  Started container hz-mc-pvc-hazelcast-mancenter
```

To check possible fix for the issue does work or not, we need to reproduce the issue. Could you check my steps and compare your commands or setup then give some ideas or more reproducer details?

Thanks in advance.

**Comment by jperucha** [comment]
Hi @hasancelik

Your response is the normal/good flow i should have gotten the first time I tried to use a pvc to store all mancenter info, including credentials as this method was used to try to workaround the other issue i opened and that it is already fixed.

I have replicated the same steps that caused the issue in first place (almost the same steps that you have taken) and found that now the issue does not appear in v3.5.0 as mancenter pod recreates without problem binding itself with the manually created PVC. Therefore I consider the issue resolved and propose to mark it as solved.

Thanks for your efforts and work towards solving it.

**Comment by erosb** [comment]
Hello, we added a system property for disabling the lock file check, [here is the documentation PR](https://github.com/hazelcast/management-center-docs/pull/36/files).

**Comment by hasancelik** [comment]
great, thanks guys 👍 @erosb @emre-aydin 

Will it be released with the upcoming release(4.2021.04)?

**Comment by erosb** [comment]
Yes, it will be in 4.2021.04.
